libcontainer/cgroups/systemd/user.go

```diff
@@ -0,0 +1,86 @@
+// +build linux
+
+package systemd
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+	dbus "github.com/godbus/dbus/v5"
+	"github.com/opencontainers/runc/libcontainer/system"
+	"github.com/pkg/errors"
+)
+
+func newUserSystemdDbus() (*systemdDbus.Conn, error) {
+	addr, err := detectUserDbusSessionBusAddress()
+	if err != nil {
+		return nil, err
+	}
+	uid, err := detectUID()
+	if err != nil {
+		return nil, err
+	}
+
+	return systemdDbus.NewConnection(func() (*dbus.Conn, error) {
+		conn, err := dbus.Dial(addr)
+		if err != nil {
+			return nil, err
+		}
+		methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(uid))}
+		err = conn.Auth(methods)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+		if err = conn.Hello(); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		return conn, nil
+	})
+}
+
+// detectUID detects UID from $XDG_RUNTIME_DIR if running in userNS.
+// Otherwise returns os.Getuid() .
+func detectUID() (int, error) {
+	if !system.RunningInUserNS() {
+		return os.Getuid(), nil
+	}
+	// xdr is like "/run/user/1000" when the UID in the initial userNS is 1000
+	xdr := os.Getenv("XDG_RUNTIME_DIR")
+	if xdr == "" {
+		return -1, errors.New("could not detect the UID from XDG_RUNTIME_DIR")
+	}
+	s := strings.TrimPrefix(filepath.Clean(xdr), "/run/user/")
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		return -1, errors.Wrap(err, "could not detect the UID from XDG_RUNTIME_DIR")
+	}
+	return i, nil
+}
+
+// detectUserDbusSessionBusAddress returns $DBUS_SESSION_BUS_ADDRESS if set.
+// Otherwise parse the value from `systemctl --user show-environment` .
+func detectUserDbusSessionBusAddress() (string, error) {
+	if env := os.Getenv("DBUS_SESSION_BUS_ADDRESS"); env != "" {
+		return env, nil
+	}
+	b, err := exec.Command("systemctl", "--user", "show-environment").CombinedOutput()
```

Comment:
If your environment variables are borked, this won't work either...

```
$ DBUS_SESSION_BUS_ADDRESS=broken XDG_RUNTIME_DIR=/broken systemctl --user show-environment
Failed to connect to bus: Connection refused
```

The code used to connect to the user bus is [here](https://github.com/systemd/systemd/blob/v245/src/libsystemd/sd-bus/sd-bus.c#L1327-L1333), it depends on either `$DBUS_SESSION_BUS_ADDRESS` or `$XDG_RUNTIME_DIR` being properly set.

My 2¢ is that this should simply fail if the `$DBUS_SESSION_BUS_ADDRESS` environment variable is not available.

Reply:
`systemctl --user show-environment` works for me with `unset DBUS_SESSION_BUS_ADDRESS`. 

Actually, podman sets `XDG_RUNTIME_DIR` but does not set `DBUS_SESSION_BUS_ADDRESS`.

Reply:
Ah I see!

The [logic in sd-bus](https://github.com/systemd/systemd/blob/v245/src/libsystemd/sd-bus/bus-internal.h#L21) falls back to parsing `$XDG_RUNTIME_DIR`, so maybe use that same logic here too?

I think that's a superior approach to spawning an external command.

(BTW, @giuseppe maybe you'd want to address the issue of Podman exporting `$XDG_RUNTIME_DIR` but not `$DBUS_SESSION_BUS_ADDRESS` if that's applicable? Maybe it's not really applicable, just pinging you here in case you think it's worth looking into...)

Reply:
updated to use `$XDG_RUNTIME_DIR/bus`